### ISSUE TITLE
Setting DNSMASQ_USER breaks status check

### ISSUE THREAD

**Issue by Beanow** [pi-hole/docker-pi-hole#789]
### Versions

  Pi-hole version is v5.2.4 (Latest: v5.2.4)
  AdminLTE version is v5.3.2 (Latest: v5.3.2)
  FTL version is v5.6 (Latest: v5.6)

### Platform
Tested on Docker 20.10.3 linux/amd64 and linux/arm64.

### Expected behavior

`pihole status` will correctly report status when `DNSMASQ_USER` is set to a non-root user.

### Actual behavior / bug

Whenever `DNSMASQ_USER` is not root `pihole status` returns
<pre>[<font color="#EF2929">✗</font>] DNS service is NOT listening</pre>

### Steps to reproduce

<details><summary><code>docker run --rm -ti <strong>-e DNSMASQ_USER=pihole</strong> pihole/pihole:v5.6 lsof -Pni:53</code></summary>

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] 01-resolver-resolv: applying... 
[fix-attrs.d] 01-resolver-resolv: exited 0.
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 20-start.sh: executing... 
 ::: Starting docker specific checks & setup for docker pihole/pihole
Assigning random password: qC2kCuXN

  [i] Installing configs from /etc/.pihole...
  [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
  [✓] Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf
chown: cannot access '': No such file or directory
chmod: cannot access '': No such file or directory
chown: cannot access '/etc/pihole/dhcp.leases': No such file or directory
Existing DNS servers detected in setupVars.conf. Leaving them alone
Setting password: qC2kCuXN
+ pihole -a -p qC2kCuXN qC2kCuXN
  [✓] New password set
DNSMasq binding to default interface: eth0
Added ENV to php:
			"PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
			"ServerIP" => "0.0.0.0",
			"VIRTUAL_HOST" => "0.0.0.0",
Using IPv4 and IPv6
::: setup_blocklists now setting default blocklists up: 
::: TIP: Use a docker volume for /etc/pihole/adlists.list if you want to customize for first boot
::: Blocklists (/etc/pihole/adlists.list) now set to:
https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
::: Testing pihole-FTL DNS: FTL started!
::: Testing lighttpd config: Syntax OK
::: All config checks passed, cleared for startup ...
::: Enabling Query Logging
  [i] Enabling logging...
  [✓] Restarting DNS server
  [✓] Logging has been enabled!
 ::: Docker start setup complete
  [i] Neutrino emissions detected...
  [✓] Pulling blocklist source list into range

  [✓] Preparing new gravity database
  [i] Using libz compression

  [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
  [✓] Status: Retrieval successful
  [i] Received 60600 domains

  [✓] Storing downloaded domains in new gravity database
  [✓] Building tree
  [✓] Swapping databases
  [i] Number of gravity domains: 60600 (60600 unique domains)
  [i] Number of exact blacklisted domains: 0
  [i] Number of regex blacklist filters: 0
  [i] Number of exact whitelisted domains: 0
  [i] Number of regex whitelist filters: 0
  [✓] Flushing DNS cache
  [✓] Cleaning up stray matter

  [✗] DNS service is NOT listening
  Pi-hole version is v5.2.4 (Latest: v5.2.4)
  AdminLTE version is v5.3.2 (Latest: v5.3.2)
  FTL version is v5.6 (Latest: v5.6)
[cont-init.d] 20-start.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
Starting pihole-FTL (no-daemon) as pihole
Starting lighttpd
[services.d] done.
Starting crond
[cmd] lsof exited 1
Stopping pihole-FTL
Stopping cron
Stopping lighttpd
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.
```
</details>

This fails and reports:
```
  [✗] DNS service is NOT listening
...
[cmd] lsof exited 1
```

<details><summary><code>docker run --rm -ti <strong>-e DNSMASQ_USER=root</strong> pihole/pihole:v5.6 lsof -Pni:53</code></summary>

```
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] 01-resolver-resolv: applying... 
[fix-attrs.d] 01-resolver-resolv: exited 0.
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 20-start.sh: executing... 
 ::: Starting docker specific checks & setup for docker pihole/pihole
Assigning random password: EG-lmxVK

  [i] Installing configs from /etc/.pihole...
  [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
  [✓] Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf
chown: cannot access '': No such file or directory
chmod: cannot access '': No such file or directory
chown: cannot access '/etc/pihole/dhcp.leases': No such file or directory
Existing DNS servers detected in setupVars.conf. Leaving them alone
Setting password: EG-lmxVK
+ pihole -a -p EG-lmxVK EG-lmxVK
  [✓] New password set
DNSMasq binding to default interface: eth0
Added ENV to php:
			"PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
			"ServerIP" => "0.0.0.0",
			"VIRTUAL_HOST" => "0.0.0.0",
Using IPv4 and IPv6
::: setup_blocklists now setting default blocklists up: 
::: TIP: Use a docker volume for /etc/pihole/adlists.list if you want to customize for first boot
::: Blocklists (/etc/pihole/adlists.list) now set to:
https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
::: Testing pihole-FTL DNS: FTL started!
::: Testing lighttpd config: Syntax OK
::: All config checks passed, cleared for startup ...
::: Enabling Query Logging
  [i] Enabling logging...
  [✓] Restarting DNS server
  [✓] Logging has been enabled!
 ::: Docker start setup complete
  [i] Neutrino emissions detected...
  [✓] Pulling blocklist source list into range

  [✓] Preparing new gravity database
  [i] Using libz compression

  [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
  [✓] Status: Retrieval successful
  [i] Received 60600 domains

  [✓] Storing downloaded domains in new gravity database
  [✓] Building tree
  [✓] Swapping databases
  [i] Number of gravity domains: 60600 (60600 unique domains)
  [i] Number of exact blacklisted domains: 0
  [i] Number of regex blacklist filters: 0
  [i] Number of exact whitelisted domains: 0
  [i] Number of regex whitelist filters: 0
  [✓] Flushing DNS cache
  [✓] Cleaning up stray matter

  [✓] DNS service is listening
     [✓] UDP (IPv4)
     [✓] TCP (IPv4)
     [✓] UDP (IPv6)
     [✓] TCP (IPv6)

  [i] Pi-hole blocking will be enabled
  [i] Enabling blocking
  [✓] Flushing DNS cache
  [✓] Pi-hole Enabled
  Pi-hole version is v5.2.4 (Latest: v5.2.4)
  AdminLTE version is v5.3.2 (Latest: v5.3.2)
  FTL version is v5.6 (Latest: v5.6)
[cont-init.d] 20-start.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
Starting pihole-FTL (no-daemon) as root
Starting crond
[services.d] done.
Starting lighttpd
COMMAND   PID USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
pihole-FT 520 root    4u  IPv4 352459      0t0  UDP *:53 
pihole-FT 520 root    5u  IPv4 352460      0t0  TCP *:53 (LISTEN)
pihole-FT 520 root    6u  IPv6 352461      0t0  UDP *:53 
pihole-FT 520 root    7u  IPv6 352462      0t0  TCP *:53 (LISTEN)
[cmd] lsof exited 0
Stopping lighttpd
Stopping cron
Stopping pihole-FTL
[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] waiting for services.
[s6-finish] sending all processes the TERM signal.
[s6-finish] sending all processes the KILL signal and exiting.
```
</details>

This works and reports:
```
  [✓] DNS service is listening
     [✓] UDP (IPv4)
     [✓] TCP (IPv4)
     [✓] UDP (IPv6)
     [✓] TCP (IPv6)
...
COMMAND   PID USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
pihole-FT 520 root    4u  IPv4 352459      0t0  UDP *:53 
pihole-FT 520 root    5u  IPv4 352460      0t0  TCP *:53 (LISTEN)
pihole-FT 520 root    6u  IPv6 352461      0t0  UDP *:53 
pihole-FT 520 root    7u  IPv6 352462      0t0  TCP *:53 (LISTEN)
[cmd] lsof exited 0
```

#### Additional context

This is probably a regression since switching to the `lsof` based detection method in https://github.com/pi-hole/pi-hole/pull/3524

**Comment by PromoFaux** [comment]
I might be entirely wrong here - but `pihole-FTL` needs to run as root in order to bind to port 53. That said, permissions and the like on linux are not a strong point for me

**Comment by dschaper** [comment]
There's two approaches (I don't remember what we did with docker.)

1. Start as root and drop priv's once the port is captured.
2. Run as non-root with a binary that has extra capabilities set.

**Comment by dschaper** [comment]
As for this, are you sure the `pihole` user exists inside the container?

**Comment by Beanow** [comment]
@dschaper yes the user exists

```sh
docker run --rm -ti --entrypoint="" pihole/pihole:v5.6 id pihole
# uid=999(pihole) gid=999(pihole) groups=999(pihole)
```

It's done as part of `automated install/basic-install.sh`
https://github.com/pi-hole/pi-hole/blob/cbfb58f7a283c2a3e7aad95a834a0287175ccb24/automated%20install/basic-install.sh#L1845

Invoked by the Dockerfile here
https://github.com/pi-hole/docker-pi-hole/blob/79ead80720225cdd649078655a6dae7caa8693e2/install.sh#L32-L34

---

But I should note this definitely looks like a regression due to the PR I mentioned before.

I used to have the `DNSMASQ_USER=pihole` variable enabled on previous versions of the docker image. For example:

```sh
docker run --rm -ti -e DNSMASQ_USER=pihole pihole/pihole:v5.1.2 lsof -Pni:53
```

```
  [✓] DNS service is running
  [i] Pi-hole blocking will be enabled
  [i] Enabling blocking
  [✓] Flushing DNS cache
  [✓] Pi-hole Enabled
  Pi-hole version is v5.1.2 (Latest: v5.1.2)
  AdminLTE version is v5.1.1 (Latest: v5.1.1)
  FTL version is v5.2 (Latest: v5.2)
...
[cmd] lsof exited 1
```

The lsof detection method didn't work in that version either, but we didn't rely on it for `pihole status`.

**Comment by R8s6** [comment]
I confirm I have exactly the same issue on a VPS running debian v10 with docker v5.7 and v5.6. 

When environment `-e DNSMASQ_USER="pihole"` is removed, the problem is gone.

My docker is running the [rootless mode](https://docs.docker.com/engine/security/rootless/), so I'm not sure if this is a factor.

However, on another ubuntu 18.04 VM with regular docker (not rootless), the `-e DNSMASQ_USER="pihole"` environment value didn't cause any issues.

**Comment by Beanow** [comment]
> My docker is running the [rootless mode](https://docs.docker.com/engine/security/rootless/), so I'm not sure if this is a factor.

I'm not running rootless, so that doesn't seem to prevent the issue here.

> However, on a Synology DS1815+ DSM with regular docker (not rootless), the same `-e DNSMASQ_USER="pihole"` didn't cause any issues.

Which version of pihole are you running there? As I noted at least `v5.1.2` of the docker image worked (as it did the TCP probe rather than lsof).

**Comment by R8s6** [comment]
> Which version of pihole are you running there? As I noted at least `v5.1.2` of the docker image worked (as it did the TCP probe rather than lsof).
On the debian VPS I tried both v5.7 and v5.6 and both had the same problem;
On Synology DSM, it's running the latest v5.7 with no issues with `-e DNSMASQ_USER="pihole"`.

**Comment by Beanow** [comment]
That's very interesting @R8s6
A quick search shows that Synology model is linux based with an x86_64 atom CPU.
So in theory you'd expect docker containers to behave very similarly.

Either way, it seems in some configurations `lsof` does work in docker and some it doesn't. :thinking:

Perhaps it's the rootless mode after all?
Perhaps there's some kernel modifications synology did?
Perhaps either one uses a different `/etc/docker/daemon.json` configuration?
If you're able to help rule some of those out, that would be awesome.

But (if I eventually have some time to try things out in some VMs), some information that would help me out would be:

- What command / compose file / stack file, you're using to test on the Synology.
- `uname -rmo` to check the Synology kernel and cpu details.
- `ps -o cmd --pid "$(pgrep dockerd)"` to check if special options are given to dockerd. Especially `--config-file`.
- Contents, if any, of `daemon.json` used (either `/etc/docker/daemon.json` by default, or the path of `--config-file`).

Cheers :smile:

**Comment by R8s6** [comment]
Oh shoot, my bad, I apologize: I checked my Synology again, it did have `DNSMASQ_USER="root"` all this time, I must have set it sometime ago and completely forgotten about it. 

As an experiment, on this Synology DS1815+, just nowI set it to `DNSMASQ_USER="pihole"`, and I got the same "DNS service not running" error;
Changing it back to `DNSMASQ_USER="root"` and restarted the docker, and the problem was gone.

TLDR: Synology is no exception, the problem is consistent (good). `DNSMASQ_USER="pihole"` appears to be the link.
